### PR TITLE
Feature/25 spline testing

### PR DESCRIPTION
**Pull request by dk1844**
### Brief explanation
 1. checks out current Spline
 2. builds spline & related docker images
 3. spawns arangodb, spline and spline-db-init via docker-compose
 4. runs prepeare set of lineage posts via Jmeter having the results to work with later.
 5. creates graphs comparing reference data with currently measured.

### Man page:
```bash
 .\run_jmetered_spline.py -h
usage: run_jmetered_spline [-h] [-n] [-b SPLINE_BRANCH]

Spline backend simple testing (jmetered)

options:
  -h, --help            show this help message and exit
  -n, --no-rebuild      if specified, spline will not be rebuilt (use when rerunning on an existing codebase) (default: False)
  -b SPLINE_BRANCH, --spline-branch SPLINE_BRANCH
                        Name of spline branch to test (default: develop)
```

### Example graph:
triangle-lineage-5-200@20reads-3ops-7attr json txt res

### CODE REVIEW

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+import argparse
+import subprocess
+import docker
+import os
+import platform
+import pandas as pd
+import matplotlib.pyplot as plt
+
+GIT_CLONE_SPLINE = "git clone https://github.com/AbsaOSS/spline.git"
+GIT_CHEKOUT_SPLINE_BRANCH = "git checkout {branch}"
```
Comment (wajda) [inline spline-testing/jmetered-backed-simple/run_jmetered_spline.py:11]: typo: CHE**C**KOUT

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+def get_mvn_by_os() -> str:
+    if platform.system() == "Windows":
+        return "mvn.cmd"
+    else:
+        return "mvn"
+
+
+def build_spline(branch):
+    print(f"Getting spline (via: '{GIT_CLONE_SPLINE}')")
+    subprocess.run(GIT_CLONE_SPLINE)
```
Comment (wajda) [inline spline-testing/jmetered-backed-simple/run_jmetered_spline.py:67]: From the [subprocess doc](https://docs.python.org/3/library/subprocess.html)

> args is required for all calls and should be a string, or a sequence of program arguments. Providing a sequence of arguments is generally preferred, as it allows the module to take care of any required escaping and quoting of arguments (e.g. to permit spaces in file names). **If passing a single string, either shell must be [True](https://docs.python.org/3/library/constants.html#True) (see below) or else the string must simply name the program to be executed without specifying any arguments.**

```suggestion
    subprocess.run(GIT_CLONE_SPLINE, shell=True)
```

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py (hunk 1 above)
Comment (dk1844) [reply spline-testing/jmetered-backed-simple/run_jmetered_spline.py]: I have removed this in the mean time.

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py [hunk 3]
```diff
@@ -87,7 +88,7 @@ def run_docker_compose():
     os.makedirs(f"{root_dir}/{RESULTS_FOLDER_NAME}", exist_ok=True)
     # --exit-code-from = reports exit code from this container
     # AND implies '--abort-on-container-exit' - will 'docker-compose down' after any container has exited
-    subprocess.run("docker-compose up --exit-code-from jmeter")
+    subprocess.run("docker-compose up --exit-code-from jmeter", shell=True)
```
Comment (wajda) [inline spline-testing/jmetered-backed-simple/run_jmetered_spline.py:91]: Every `subpricess.run()` should propagate errors. What about setting `check=True` attribute for every `run()` call?

From the [doc](https://docs.python.org/3/library/subprocess.html):
> If _check_ is true, and the process exits with a non-zero exit code, a [CalledProcessError](https://docs.python.org/3/library/subprocess.html#subprocess.CalledProcessError) exception will be raised. Attributes of that exception hold the arguments, the exit code, and stdout and stderr if they were captured.

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py (hunk 2 above)
Comment (dk1844) [reply spline-testing/jmetered-backed-simple/run_jmetered_spline.py]: Thanks, applied.

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py (hunk 3 above)
Comment (dk1844) [reply spline-testing/jmetered-backed-simple/run_jmetered_spline.py]: Thanks, applied.

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env python3
+import argparse
+import subprocess
+import docker
+import os
+import sys
+import platform
+import pandas as pd
+import matplotlib.pyplot as plt
+import git
```
Comment (wajda) [inline spline-testing/jmetered-backed-simple/run_jmetered_spline.py:10]: missing library

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env python3
+import argparse
+import subprocess
+import docker
+import os
+import sys
```
Comment (wajda) [inline spline-testing/jmetered-backed-simple/run_jmetered_spline.py:6]: unused import

**Comment by wajda** [comment]
Once the Spline repo is cloned the attempt to change branch for subsequent script execution with the `-b` parameter does not have any effect.

Scenario:
1. `python run_jmetered_spline.py -b my_branch`
2. The script correctly runs successfully against **my_branch** branch
3. `python run_jmetered_spline.py -b another_branch`
4. The script is expected to run against **another_branch**, but it still runs against the **my_branch**
5. `python run_jmetered_spline.py -b not_existing_branch`
6. The script is expected to fail, but it still runs successfully against the **my_branch**

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py (hunk 5 above)
Comment (dk1844) [reply spline-testing/jmetered-backed-simple/run_jmetered_spline.py]: Thanks, fixed.

File: spline-testing/jmetered-backed-simple/run_jmetered_spline.py (hunk 4 above)
Comment (dk1844) [reply spline-testing/jmetered-backed-simple/run_jmetered_spline.py:9]: Thanks. fixed in `requirements.txt`.

**Comment by dk1844** [comment]
> Every time I run the script I get the following error:
> 
> image
> 
> The error is recoverable, but requires user interaction (I have to manually confirm `Y` in the console)

Having replaced Rancher with the official Docker Desktop on my machine, I was able to reproduce the issue.

According to https://github.com/docker/compose/issues/7933, the message is possibly misleading. I have renamed the images a bit to conform to the format `registry/namespace/repository:tag` which seemed to have solved the issue on my machine. I'd be glad if you could confirm on your end as well.

**Comment by wajda** [comment]
Now I've got this:

```
❯ python run_jmetered_spline.py -n                           
Traceback (most recent call last):
  File "/usr/bin/docker-compose", line 33, in <module>
    sys.exit(load_entry_point('docker-compose==1.29.2', 'console_scripts', 'docker-compose')())
  File "/usr/bin/docker-compose", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/usr/lib64/python3.10/importlib/metadata/__init__.py", line 171, in load
    module = import_module(match.group('module'))
  File "/usr/lib64/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/lib/python3.10/site-packages/compose/cli/main.py", line 29, in <module>
    from ..metrics.decorator import metrics
  File "/usr/lib/python3.10/site-packages/compose/metrics/decorator.py", line 3, in <module>
    from compose.metrics.client import MetricsCommand
  File "/usr/lib/python3.10/site-packages/compose/metrics/client.py", line 5, in <module>
    from docker import ContextAPI
ImportError: cannot import name 'ContextAPI' from 'docker' (/home/wajda/.local/lib/python3.10/site-packages/docker/__init__.py)
docker-compose up done
Generating graph for file processed_results/chain-lineage-5reads-1-501@50ops-4attr.json.txt.res
  Using variable column 'operationCount' (nuniques = 11)
QSocketNotifier: Can only be used with threads started with QThread
Generating graph for file processed_results/triangle-lineage-5reads-3-10@10ops-4attr.json.txt.res
  Fall-backing variable column to 'timeStamp'
Generating graph for file processed_results/triangle-lineage-5reads-3ops-1-501@50attr.json.txt.res
  Using variable column 'attributeCount' (nuniques = 11)
Generating graph for file processed_results/chain-lineage-5reads-3ops-1-501@50attr.json.txt.res
  Using variable column 'attributeCount' (nuniques = 11)
Generating graph for file processed_results/diamond-lineage-5reads-3ops-1-501@50attr.json.txt.res
  Using variable column 'attributeCount' (nuniques = 11)
Generating graph for file processed_results/triangle-lineage-5reads-1-501@50ops-4attr.json.txt.res
  Using variable column 'operationCount' (nuniques = 11)
Generating graph for file processed_results/diamond-lineage-5reads-1-501@50ops-4attr.json.txt.res
  Using variable column 'operationCount' (nuniques = 11)
Generating graph for file processed_results/triangle-lineage-1-501@50reads-3ops-4attr.json.txt.res
  Using variable column 'readCount' (nuniques = 11)
Generating graph for file processed_results/chain-lineage-1-2001@50reads-3ops-4attr.json.txt.res
  Using variable column 'readCount' (nuniques = 41)
Generating graph for file processed_results/diamond-lineage-1-1001@50reads-3ops-4attr.json.txt.res
  Using variable column 'readCount' (nuniques = 21)
Cleaning up custom image 'local-only/spline-testing/testing-spline-db-admin:1.0.0-SNAPSHOT'
Traceback (most recent call last):
  File "/home/wajda/Projects/spline-getting-started/spline-testing/jmetered-backed-simple/run_jmetered_spline.py", line 104, in cleanup_docker
    client.images.remove(image_name, force=True)
AttributeError: 'function' object has no attribute 'remove'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/wajda/Projects/spline-getting-started/spline-testing/jmetered-backed-simple/run_jmetered_spline.py", line 244, in <module>
    run(args)
  File "/home/wajda/Projects/spline-getting-started/spline-testing/jmetered-backed-simple/run_jmetered_spline.py", line 233, in run
    cleanup_docker()
  File "/home/wajda/Projects/spline-getting-started/spline-testing/jmetered-backed-simple/run_jmetered_spline.py", line 106, in cleanup_docker
    except docker.errors.ImageNotFound as inf:
AttributeError: module 'docker.errors' has no attribute 'ImageNotFound'

```

**Review by wajda — APPROVED** [review]
LGTM

**Pull request merged**